library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/VideoDecodeThread.kt

```diff
@@ -33,52 +35,46 @@ class VideoDecodeThread (
             Pair(
-                Util.ceilDivide(width, widthAlignment) * widthAlignment,
-                Util.ceilDivide(height, heightAlignment) * heightAlignment)
+                    Util.ceilDivide(width, widthAlignment) * widthAlignment,
+                    Util.ceilDivide(height, heightAlignment) * heightAlignment)
         }
     }
 
     override fun run() {
         if (DEBUG) Log.d(TAG, "$name started")
 
-        val decoder = MediaCodec.createDecoderByType(mimeType)
-        val widthHeight = getDecoderSafeWidthHeight(decoder)
-        val format = MediaFormat.createVideoFormat(mimeType, widthHeight.first, widthHeight.second)
+        try {
+            val decoder = MediaCodec.createDecoderByType(mimeType)
+            val widthHeight = getDecoderSafeWidthHeight(decoder)
+            val format = MediaFormat.createVideoFormat(mimeType, widthHeight.first, widthHeight.second)
 
-        decoder.setOnFrameRenderedListener(onFrameRenderedListener, null)
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                decoder.setOnFrameRenderedListener(onFrameRenderedListener, null)
+            }
 
-        if (DEBUG) Log.d(TAG, "Configuring surface ${widthHeight.first}x${widthHeight.second} w/ '$mimeType', max instances: ${decoder.codecInfo.getCapabilitiesForType(mimeType).maxSupportedInstances}")
-        try {
+            if (DEBUG) Log.d(TAG, "Configuring surface ${widthHeight.first}x${widthHeight.second} w/ '$mimeType', max instances: ${decoder.codecInfo.getCapabilitiesForType(mimeType).maxSupportedInstances}")
             decoder.configure(format, surface, null, 0)
-        } catch (e: IllegalArgumentException) {
-            if (DEBUG) Log.d(TAG, "$name stopped due to '${e.message}'")
-            // While configuring stopAsync can be called and surface released. Just exit.
-            if (isRunning) e.printStackTrace()
-            return
-        }
-        decoder.start()
-        if (DEBUG) Log.d(TAG, "Started surface decoder")
 
-        val bufferInfo = MediaCodec.BufferInfo()
+            // TODO: add scale option (ie: FIT, SCALE_CROP, SCALE_NO_CROP)
+            //decoder.setVideoScalingMode(MediaCodec.VIDEO_SCALING_MODE_SCALE_TO_FIT_WITH_CROPPING);
 
-        // Main loop
-        while (isRunning) {
-            val inIndex: Int = decoder.dequeueInputBuffer(10000L)
-            if (inIndex >= 0) {
-                // fill inputBuffers[inputBufferIndex] with valid data
-                var byteBuffer: ByteBuffer?
-                try {
+            decoder.start()
+            if (DEBUG) Log.d(TAG, "Started surface decoder")
+
+            val bufferInfo = MediaCodec.BufferInfo()
+
+            // Main loop
+            while (!exitFlag.get()) {
+                val inIndex: Int = decoder.dequeueInputBuffer(10000L)
+                if (inIndex >= 0) {
+                    // fill inputBuffers[inputBufferIndex] with valid data
+                    var byteBuffer: ByteBuffer?
```

Comment:
val byteBuffer: ByteBuffer? = decoder.getInputBuffer(inIndex)

Reply:
> BTW: You changed `boolean isRunning = true` to `exitFlag: AtomicBoolean = AtomicBoolean(false)` What is the reason for that? Do not understand.

Hi @alexeyvasilyev - when testing I noticed sometimes when the screen turned off the video decode thread kept on going.. I noticed this because the logs were showing this line over and over:
```
"Cannot add frame, queue is full"
```

I verified stopAsync() was being called. However, I did some debugging and noticed that `isRunning` flag was being accessed from 2 different threads. I figured either the AtomicBoolean would work or to wrap access to isRunning in some synchronized block. Or maybe even using volatile on the isRunning variable.

Anyway, after that change I didn't see the log message printed anymore once the screen turned off

Reply:
Thanks for explanation!